source/module_base/test/complexarray_test.cpp

```diff
@@ -0,0 +1,207 @@
+*         value to first ComplexArray
+*       - throw error when one ComplexArray minus another ComplexArray
+*         that has different dimension.
+*
+*   - operator "*":
+*       - one ComplexArray is multiplied by a double
+*       - one ComplexArray is multiplied by a complex
+*       - one ComplexArray is multiplied by another ComplexArray that has same dimension
+*       - throw error when one ComplexArray is miltiplied by another ComplexArray
+*         that has different dimension.
+*
+*   - operator "*=":
+*       similar as "*"
+*
+*   - oprator "()":
+*       - assess the element
+*/
+
+
+class ComplexArray_test : public testing::Test
+{
+    protected:
+        ModuleBase::ComplexArray a2,a4,b2,b4,c2,c4,d2;
+        ModuleBase::ComplexArray a2_plus_b2,a2_minus_b2,a2_mult_b2,a2_mult_3;
+        std::complex<double> com1 {1.0,2.0};
+        std::complex<double> com2 {3.0,4.0};
+
+        void SetUp()
+        {
+            a2 = ModuleBase::ComplexArray(2,1,1,1);  // if define this class as a matrix
+            b2 = ModuleBase::ComplexArray(2,1,1,1);  // of 4 dimesions,
+            c2 = ModuleBase::ComplexArray(2,1,1,1);  // is a2 +/-/* d2 allowed or not???
+            d2 = ModuleBase::ComplexArray(1,2,1,1);  // it does not matter, this situation will not appear in ABACUS
+            a2_plus_b2 = ModuleBase::ComplexArray(2,1,1,1);
+            a2_minus_b2 = ModuleBase::ComplexArray(2,1,1,1);
+            a2_mult_b2 = ModuleBase::ComplexArray(2,1,1,1);
+            a2_mult_3 = ModuleBase::ComplexArray(2,1,1,1);
+            a4 = ModuleBase::ComplexArray(2,2,1,1);
+            b4 = ModuleBase::ComplexArray(2,2,1,1);
+            c4 = ModuleBase::ComplexArray(2,2,1,1);
+            a2 = com1;
+            b2 = com2;
+            a4 = com1;
+            b4 = com2;
+            d2 = com1;
+            a2_plus_b2 = com1 + com2;
+            a2_minus_b2 = com1 - com2;
+            a2_mult_b2 = com1 * com2;
+            a2_mult_3 = com1 * 3.0;
+        }
+
+};
+
+bool Compare(ModuleBase::ComplexArray &ca1, ModuleBase::ComplexArray &ca2)
+{
+    if (ca1.getSize() != ca2.getSize()) {return false;}
+    if (ca1.getBound1() != ca2.getBound1()) {return false;}
+    if (ca1.getBound2() != ca2.getBound2()) {return false;}
+    if (ca1.getBound3() != ca2.getBound3()) {return false;}
+    if (ca1.getBound4() != ca2.getBound4()) {return false;}
+    for ( int i = 0;i < ca1.getBound1();i++)
+    {
+        for ( int j = 0;j < ca1.getBound2();j++)
+        {
+            for ( int k = 0;k < ca1.getBound3();k++)
+            {
+                for ( int l = 0;l < ca1.getBound4();l++)
+                {
+                    if (ca1(i,j,k,l) != ca2(i,j,k,l)) {return false;}
+                }
+            }
+        }
+    }
+    return true;
+}
+
+
+TEST_F(ComplexArray_test,operator_equal)
+{   
+    c2 = a2;
+    b2 = a2;
+    EXPECT_TRUE(Compare(c2,a2));
+    EXPECT_TRUE(Compare(b2,a2));
```

Comment:
What is the purpose to test twice?

Reply:
b2 is assigned by a complex, while c2 is not assigned. These should be two different situations.

Reply:
> b2 is assigned by a complex, while c2 is not assigned. These should be two different situations.

Hi @pxlxingliang,
I don't think they are different situations, both cases are calling the same instructions. 
https://github.com/deepmodeling/abacus-develop/blob/9aa165087b1049eb668165a9c6e354336374e903/source/module_base/complexmatrix.cpp#L226-L231
Am I right? Or I've ignored something? Whatever, adding a comment explaining it is a good practice.

However, you did not test this function:
https://github.com/deepmodeling/abacus-develop/blob/9aa165087b1049eb668165a9c6e354336374e903/source/module_base/complexmatrix.cpp#L234-L242

Reply:
this function is actually not used in ABACUS, i think the test is unnecessary